File: cpp/letterCombinations.cpp
#include <cstdlib>
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

template <class T>
void printVector(vector<T> vec){
    for (T v : vec){
        cout << v << ',' ;
    }
    cout << endl;
}

vector < string > letterCombinations(string digits){
    unordered_map<int, string> m = {{2, "abc"}, {3, "def"},{4, "ghi"}, {5, "jkl"},{6, "mno"},{7, "pqrs"},{8, "tuv"},{9, "wxyz"}};
    // vector<string> old = {""};
    vector<string> old = {""};
    // vector<string> next = {""};
    vector<string> next ;

    if(digits.empty()){
        return {};
    }

    for(char c : digits){
        cout << m[c-'0'] << endl;
        // printVector(old);
        for(char d : m[c-'0']){
            for (string o : old){
                next.push_back(o+d);
            }
        }
        old = next;
        next = {};
    }

    return old;
}



int main(){

    string digits = "23";
    // digits = "2";
    digits = "";

    printVector(letterCombinations(digits));
    // cout << letterCombinations(digits) << endl;

    return 0;
}

File: python/letterCombinations.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        letters = [['a', 'b', 'c'],
        ['d', 'e', 'f'],
        ['g', 'h', 'i'],
        ['j', 'k', 'l'],
        ['m', 'n', 'o'],
        ['p', 'q', 'r', 's'],
        ['t', 'u', 'v'],
        ['w', 'x', 'y', 'z']]
        c = []
        # print (len(digits) )
        if len(digits)== 0:
            return []
        elif len(digits) > 1:
            # print (digits[0:-1])
            c = self.letterCombinations(digits[0:-1])
            # print ('c',c)
        else:
            # print ( int(digits[-1]) -2 )
            # print (letters[ int(digits[-1]) -2 ])
            return letters[ int(digits[-1]) -2 ]

        out = []
        for i in c:
            # print ( int(digits[-1]) -2 , letters[ int(digits[-1]) -2 ] )
            for j in letters[ int(digits[-1]) -2 ] :
                out.append(i + j)
        # print (a)
        # print (b)
        # print (c)
        return out

s = Solution()
# s.letterCombinations('2')
# o = s.letterCombinations('23')
# o = s.letterCombinations('92')
o = s.letterCombinations('')
print(o)
